lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,309 @@
+     * @param {Node} node
+     */
+    function collapse(node) {
+      while (node.children && node.children.length === 1) {
+        node.name += '/' + node.children[0].name;
+        node.children = node.children[0].children;
+      }
+
+      if (node.children) {
+        for (const child of node.children) {
+          collapse(child);
+        }
+      }
+    }
+    collapse(rootNode);
+
+    // TODO(cjamcl): Should this structure be flattened for space savings?
+    // Like DOM Snapshot.
+    // Less JSON (no super nested children, and no repeated property names).
+
+    return rootNode;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<RootNode[]>}
+   */
+  static async makeJavaScriptRootNodes(artifacts, context) {
+    /** @type {RootNode[]} */
+    const rootNodes = [];
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplication = await ModuleDuplication.request(artifacts, context);
+
+    /** @type {Array<{src: string, length: number, unusedJavascriptSummary?: import('../computed/unused-javascript-summary.js').Summary}>} */
+    const scriptData = [];
+    const inlineScriptData = {
+      src: artifacts.URL.finalUrl,
+      length: 0,
+    };
+    for (const scriptElement of artifacts.ScriptElements) {
+      // Normalize ScriptElements so that inline scripts show up as a single entity.
+      if (!scriptElement.src) {
+        inlineScriptData.length += (scriptElement.content || '').length;
+        continue;
+      }
+
+      const url = scriptElement.src;
+      const bundle = bundles.find(bundle => url === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[url];
+      if (!bundle || !scriptCoverages) continue;
+
+      scriptData.push({
+        src: scriptElement.src,
+        length: (scriptElement.content || '').length,
+        unusedJavascriptSummary:
+          await UnusedJavaScriptSummary.request({url, scriptCoverages, bundle}, context),
+      });
+    }
+    if (inlineScriptData.length) scriptData.unshift(inlineScriptData);
+
+    for (const {src, length, unusedJavascriptSummary} of scriptData) {
+      const bundle = bundles.find(bundle => bundle.script.src === src);
+      const name = src;
+
+      let node;
+      if (bundle && unusedJavascriptSummary && unusedJavascriptSummary.sourcesWastedBytes) {
+        /** @type {Record<string, SourceData>} */
+        const sourcesData = {};
+        for (const source of Object.keys(bundle.sizes.files)) {
+          /** @type {SourceData} */
+          const sourceData = {
+            resourceBytes: bundle.sizes.files[source],
+          };
+
+          if (unusedJavascriptSummary && unusedJavascriptSummary.sourcesWastedBytes) {
+            sourceData.unusedBytes = unusedJavascriptSummary.sourcesWastedBytes[source];
+          }
+
+          if (duplication) {
+            const key = ModuleDuplication._normalizeSource(source);
```

Comment:
we should drop the leading `_` on the function name if it's not going to be private to the class